packages/eslint-plugin/tests/rules/no-base-to-string.test.ts

```diff
@@ -353,5 +380,190 @@ String(...objects);
         },
       ],
     },
+
+    {
+      code: `
+        [{}, {}].join('');
+      `,
+      errors: [
+        {
+          data: {
+            certainty: 'will',
+            name: '[{}, {}]',
+          },
+          messageId: 'baseArrayJoin',
+        },
+      ],
+    },
+    {
+      code: `
+        class A {}
+        [{}, 'str'].join('');
+      `,
+      errors: [
+        {
+          data: {
+            certainty: 'may',
```

Comment:
This is a bit obsessive of me, but I think we really need this to be `will`. 

Here's what I was playing around with...

[Playground](https://deploy-preview-10287--typescript-eslint.netlify.app/play/#ts=5.6.2&fileType=.ts&code=PTAEAcCcHsDcEMBGAbAnhSBLAdgFwM6gDkA2tIgFYCmAxrqAPKW24C6RAdKAMoAW0AV2QATUIirEA7pmTJOAKBIBvAL6sOFaDgAURADTEAlAG558kBjhI0GHAWJlmdRk7YGAZtGice-IaPEpGTkORVUDIk9vdU0dfSNTczAAYUERWzxiKI8veOhIB3JqZyZit1Ao%2BKpcGi4%2BNICJIgBbeFQFYVpkeEgJGmhsfHpW1HFSgC5QVVAAHyzc0xIRscoDZapSmK1sXQiTUFALAB4AWhPQAEkW0Cp8cFpMeFl0fAFIKEx8KlFcXk-QAZ9N69PA2fBtQgAImkskhZgs4OaEnghCgVFgmEE%2BAMVAAHjQqOB6D1EJhcJAeuhkFRsABzX51PzpQItNodLo9PoDIagHqQSYAQUgFNQR1wqHu0HcoHWpQAfPI%2BRptrsEkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oCMBDZRLXxdk%2BKkwDm6KAHde0JpHBgAviBVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

```ts
// provably prints '[object Object]'. Should be 'will'.
[{}].join(', ');

// provably prints '[object Object], foo'. Should be 'will'.
[{}, 'foo'].join(', ');

// Could print 'foo, foo', or '[object Object], foo', etc. Should be 'may'.
declare const maybeObj: {} | 'foo';
[maybeObj, maybeObj].join(', ');  // <-- I'm especially surprised this one currently says "will"

// same as previous, except arbitrary length. Should be 'may'.
declare const arr: Array<typeof maybeObj>
arr.join(', ');
```

*** 

In other words, the certainty of a tuple should be the max certainty of any of its elements, not the same as the certainty of an array whose element type is the union of the tuple's elements.

Reply:
@kirkwaiblinger 
> // provably prints '[object Object], foo'. Should be 'will'.
[{}, 'foo'].join(', ');

It would be possible to implement this if we got the type of each of the array elements. Perhaps it would be better to report on the individual elements if they are literal arrays. What do you think?

```ts
[{}, 'foo'].join(', ');
~~~~~~~~~~ // as-s

[{}, 'foo'].join(',');
 ~~ / to-be
```